### PR TITLE
Bb query base linehaul rate

### PR DESCRIPTION
**Pull request by breanneboland**
## Description

This adds the completed query (with date!) to find a base linehaul rate. This also required a redoing of a number of tests and the expansion of a couple, so as to keep things in proper working order. 

## Reviewer Notes

General workingness is plenty.

## Code Review Verification Steps

* [ ] All tests pass.
* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/blob/master/docs/backend.md#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/blob/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.
* [ ] Request review from a member of a different team.
* [ ] (TRIAL) Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/156455456) for this change

### CODE REVIEW

File: pkg/models/tariff400ng_linehaul_rate.go [hunk 1]
```diff
@@ -67,3 +67,26 @@ func (t *Tariff400ngLinehaulRate) ValidateCreate(tx *pop.Connection) (*validate.
...
+	FROM
+		tariff400ng_linehaul_rates
+	WHERE
+		(distance_miles_lower <= $1 OR distance_miles_upper > $1)
+	AND
+		(weight_lbs_lower <= $2 OR weight_lbs_upper > $2)
+	AND
+		type = $3
+	AND
+		(effective_date_lower <= $4 OR effective_date_upper > $4);`
```
Comment (kilbergr) [inline pkg/models/tariff400ng_linehaul_rate.go]: ditto.

File: pkg/models/tariff400ng_linehaul_rate.go [hunk 2]
```diff
@@ -67,3 +67,26 @@ func (t *Tariff400ngLinehaulRate) ValidateCreate(tx *pop.Connection) (*validate.
...
+
+	err = tx.RawQuery(sql, mileage, (cwt * 100), moveType, date).First(&linehaulRate)
```
Comment (kilbergr) [inline pkg/models/tariff400ng_linehaul_rate.go]: nice accounting for cwt.

File: pkg/models/tariff400ng_linehaul_rate_test.go [hunk 3]
```diff
@@ -109,3 +110,38 @@ func (suite *ModelSuite) Test_LinehaulDistanceValidation() {
 	}
 	suite.verifyValidationErrors(&invalidLinehaulRate, expErrors)
 }
+
+func (suite *ModelSuite) Test_LinehaulRateCreateAndSave() {
```
Comment (kilbergr) [inline pkg/models/tariff400ng_linehaul_rate_test.go]: I dunno what this is testing--it's fine to leave in but I think it's just testing that mustSave works. What was your intention?

File: pkg/rateengine/linehaul.go [hunk 4]
```diff
@@ -5,14 +5,17 @@ import (
 	"fmt"
 	"time"
 
+	// "github.com/gobuffalo/pop"
```
Comment (kilbergr) [inline pkg/rateengine/linehaul.go]: remove?

File: pkg/rateengine/linehaul.go [hunk 5]
```diff
@@ -23,15 +26,12 @@ func (re *RateEngine) determineMileage(originZip int, destinationZip int) (milea
...
 // Determine the Base Linehaul (BLH)
-func (re *RateEngine) baseLinehaul(mileage int, cwt int) (baseLinehaulChargeCents int, err error) {
-	// TODO (Rebecca): This will come from a fetch
-	baseLinehaulChargeCents = mileage * cwt
-	// TODO (Rebecca): make a proper error
-	if baseLinehaulChargeCents == 0 {
-		err = errors.New("Oops determineBaseLinehaul")
-	} else {
-		err = nil
+func (re *RateEngine) baseLinehaul(mileage int, cwt int, date time.Time) (baseLinehaulChargeCents int, err error) {
```
Comment (kilbergr) [inline pkg/rateengine/linehaul.go]: great.

File: pkg/rateengine/linehaul_test.go [hunk 6]
```diff
@@ -23,13 +23,32 @@ func (suite *RateEngineSuite) Test_CheckDetermineMileage() {
...
+	}
+
+	mileage := 3200
+	cwt := 39
+	date := testdatagen.DateInsidePeakRateCycle
+
+	blh, err := engine.baseLinehaul(mileage, cwt, date)
+	if blh != expected {
+		t.Errorf("BaseLinehaulCents should have been %d but is %d.", expected, blh)
 	}
```
Comment (kilbergr) [inline pkg/rateengine/linehaul_test.go]: I might add a test with more than 1 linehaul entry to ensure I returned the expected baselinehaul due to discussion above.
Something like 
```
func (suite *RateEngineSuite) Test_CheckBaseLinehaul() {
	t := suite.T()
	engine := NewRateEngine(suite.db, suite.logger)

	expected := 128000

	newBaseLinehaul := models.Tariff400ngLinehaulRate{
		DistanceMilesLower: 3101,
		DistanceMilesUpper: 3300,
		WeightLbsLower:     3000,
		WeightLbsUpper:     4000,
		RateCents:          expected,
		Type:               "ConusLinehaul",
		EffectiveDateLower: testdatagen.PeakRateCycleStart,
		EffectiveDateUpper: testdatagen.PeakRateCycleEnd,
	}

	otherBaseLinehaul := models.Tariff400ngLinehaulRate{
		DistanceMilesLower: 3401,
		DistanceMilesUpper: 3500,
		WeightLbsLower:     3000,
		WeightLbsUpper:     4000,
		RateCents:          158000,
		Type:               "ConusLinehaul",
		EffectiveDateLower: testdatagen.PeakRateCycleStart,
		EffectiveDateUpper: testdatagen.PeakRateCycleEnd,
	}
	_, err := suite.db.ValidateAndSave(&newBaseLinehaul)
	_, otherErr := suite.db.ValidateAndSave(&otherBaseLinehaul)
	if err != nil {
		t.Errorf("The object didn't save: %s", err)
	}
	if otherErr != nil {
		t.Errorf("The object didn't save: %s", err)
	}

	mileage := 3200
	cwt := 39
	date := testdatagen.DateInsidePeakRateCycle

	blh, err := engine.baseLinehaul(mileage, cwt, date)
	if blh != expected {
		t.Errorf("BaseLinehaulCents should have been %d but is %d.", expected, blh)
	}
}
```

**Comment by kilbergr** [comment]
Check out how Alexi handles making sure he returns the right number of rates. I think applying this to your query would be helpful too. https://github.com/transcom/mymove/pull/356/files#diff-61c4416001b6fecce222e70807d6d274R81

File: pkg/models/tariff400ng_linehaul_rate.go [hunk 7]
```diff
@@ -67,3 +67,26 @@ func (t *Tariff400ngLinehaulRate) ValidateCreate(tx *pop.Connection) (*validate.
...
+// FetchBaseLinehaulRate takes a move's distance and weight and queries the tariff400ng_linehaul_rates table to find a move's base linehaul rate.
+func FetchBaseLinehaulRate(tx *pop.Connection, mileage int, cwt int, date time.Time) (linehaulRate int, err error) {
+	moveType := "ConusLinehaul" // TODO: change to a parameter once we're serving more move types
+
+	sql := `SELECT
+		rate_cents
+	FROM
+		tariff400ng_linehaul_rates
+	WHERE
+		(distance_miles_lower <= $1 OR distance_miles_upper > $1)
```
[the remark this answers is not in the document]
Comment (breanneboland) [reply pkg/models/tariff400ng_linehaul_rate.go]: nyerp nyerp nyerp

An excellent catch, thank you!

File: pkg/models/tariff400ng_linehaul_rate_test.go (hunk 3 above)
Comment (breanneboland) [reply pkg/models/tariff400ng_linehaul_rate_test.go]: It was that. When Alexi and I were pairing, we were experiencing a weird glitch, and we weren't sure this was working. So we figured we'd add a test that guaranteed it, just to answer yesterday's question and address any future questions about it.

File: pkg/rateengine/linehaul_test.go (hunk 6 above)
Comment (breanneboland) [reply pkg/rateengine/linehaul_test.go]: That adds some excellent nuance. Thank you.

File: pkg/models/tariff400ng_linehaul_rate.go [hunk 8]
```diff
@@ -67,3 +68,34 @@ func (t *Tariff400ngLinehaulRate) ValidateCreate(tx *pop.Connection) (*validate.
...
+		(weight_lbs_lower <= $2 AND weight_lbs_upper > $2)
+	AND
+		type = $3
+	AND
+		(effective_date_lower <= $4 AND effective_date_upper >= $4);`
+
+	err = tx.RawQuery(sql, mileage, (cwt * 100), moveType, date).All(&linehaulRates)
+
+	if err != nil {
+		return 0, fmt.Errorf("Error fetching linehaul rate: %s", err)
```
Comment (kilbergr) [inline pkg/models/tariff400ng_linehaul_rate.go]: thanks

**Pull request merged**